### ISSUE TITLE
Open some events for external subscription

### ISSUE THREAD

**Issue by hyzx86** [sebastienros/yessql#547]
Currently we have some events open in OC, but those are not enough, some events may be an intermediate process, 

for example, we get an object and try UpdateAsync first, the object state is then saved to the state manager.
Then call some validation function ValidateAsync in OC
Then publish it

https://github.com/OrchardCMS/OrchardCore/blob/dcf83bca479ad06eeb41b9a09292a67fa26a2b21/src/OrchardCore.Modules/OrchardCore.Contents/Endpoints/Api/CreateEndpoint.cs#L59-L120

We cannot tell if a Save is the last one to be committed to the database

So my suggestion is to expose some events externally through YesSql's series of IIndexCmmand types
Used to notify external programs that the object has been committed to the database,
You can perform subsequent operations

For example, update the Lucene index and create a custom index table (this is what I am working on 😅 )

**Comment by hyzx86** [comment]
Hi @Skrypt , There's a similar problem here, right? An index action may be performed multiple times on the same document when updating a piece of content.

https://github.com/OrchardCMS/OrchardCore/blob/dcf83bca479ad06eeb41b9a09292a67fa26a2b21/src/OrchardCore.Modules/OrchardCore.Search.Lucene/Handler/LuceneIndexingContentHandler.cs#L29-L33

**Comment by Skrypt** [comment]
Good question. Normally it should not. I never saw this execute twice yet. Maybe something has changed. If we index multiple times then it is a strong performance issue.

**Comment by hyzx86** [comment]
Yes, it's true that it hasn't been updated twice so far, because Yessql's SaveAsync just adds the object to the pending collection.

I've been looking at Yessql's code over and over again, and I think it's worth opening up some of the functionality to customisation.

If I want to add some custom actions before executing the Add, Delete, and Update commands to the database, what should I do?

**Comment by hyzx86** [comment]
@Skrypt I created a branch , on my own repository, because our current project is mainly based on oc 1.8.3 It depends on yessql 4.0.
So I created this PR based on yessql 4.0 
https://github.com/hyzx86/yessql/pull/1
